### PR TITLE
Add more detail docs about using `querier.max-concurrent`.

### PR DESCRIPTION
**Pull request by kavirajk**
**What this PR does / why we need it**:
Add more detail docs about using `querier.max-concurrent` option in the querier.

The information is stolen from cortex docs
https://cortexmetrics.io/docs/configuration/arguments/#querier

**Which issue(s) this PR fixes**:
Fixes NA

**Special notes for your reviewer**:
Motivation - https://community.grafana.com/t/loki-total-worker-concurrency-is-greater-than-promql-max-concurrency/49283

**Checklist**
- [x] Documentation added
- [ ] Tests updated

### CODE REVIEW

File: docs/sources/configuration/_index.md [hunk 1]
```diff
@@ -260,7 +260,10 @@ The `querier_config` block configures the Loki Querier.
 # CLI flag: -querier.query-ingesters-within
 [query_ingesters_within: <duration> | default = 0s]
 
-# The maximum number of concurrent queries allowed.
+# The maximum number of top-level LogQL queries that will execute at the same time, per querier process.
+# If using the query frontend, this should be set to at least (`-querier.worker-parallelism` * `number of query frontend replicas`).
+# Otherwise queries may queue in the queriers and not the frontend, which will affect QoS.
```
Comment (slim-bean) [inline docs/sources/configuration/_index.md:265]: This isn't quite correct. Really it should be set to match the number of CPU cores available on the querier (or slightly less than the number of cores if you are in a shared environment like we are)

For example: running one querier with 40 worker-parallelism won't gain you much if you only have 4 cores

Ideally it should be set to match the machine cores and then you should run enough queriers to match (worker-parallelism * frontend replicas)/max-concurrent

File: pkg/querier/querier.go [hunk 2]
```diff
@@ -56,7 +56,7 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.DurationVar(&cfg.QueryTimeout, "querier.query-timeout", 1*time.Minute, "Timeout when querying backends (ingesters or storage) during the execution of a query request")
 	f.DurationVar(&cfg.ExtraQueryDelay, "querier.extra-query-delay", 0, "Time to wait before sending more than the minimum successful query requests.")
 	f.DurationVar(&cfg.QueryIngestersWithin, "querier.query-ingesters-within", 0, "Maximum lookback beyond which queries are not sent to ingester. 0 means all queries are sent to ingester.")
-	f.IntVar(&cfg.MaxConcurrent, "querier.max-concurrent", 20, "The maximum number of concurrent queries.")
+	f.IntVar(&cfg.MaxConcurrent, "querier.max-concurrent", 20, `The maximum number of top-level LogQL queries that will execute at the same time, per querier process. If using the query frontend, this should be set to at least ('-querier.worker-parallelism' * 'number of query frontend replicas'). Otherwise queries may queue in the queriers and not the frontend, which will affect QoS. Alternatively, consider using -querier.worker-match-max-concurrent to force worker parallelism to match '-querier.max-concurrent'.`)
```
Comment (slim-bean) [inline pkg/querier/querier.go:59]: Same comment as above we should update this to match the other description

File: docs/sources/configuration/_index.md [hunk 3]
```diff
@@ -260,7 +260,11 @@ The `querier_config` block configures the Loki Querier.
 # CLI flag: -querier.query-ingesters-within
 [query_ingesters_within: <duration> | default = 0s]
 
-# The maximum number of concurrent queries allowed.
+# The maximum number of top-level LogQL queries that will execute at the same time, per querier process.
+# It should be set to match the number of CPU cores available on the querier (or slightly less than the number of cores if you are in a shared environment).
+# If using the query frontend, Then you should run enough queriers to match (`-querier.worker-parallelism` * `number of query frontend replicas`)/`-querier.max-concurrent`.
+# Otherwise queries may queue in the query frontend, which will affect QoS.
```
Comment (KMiller-Grafana) [inline docs/sources/configuration/_index.md:266]: This suggestion is meant to improve readability and grammar.  I removed hyphens from the formula given for quantity of queriers.  They looked like negative signs to me, and since there was one in the numerator and one in the denominator which cancel each other out. I don't think you meant them as negative signs anyway.

```suggestion
# Set it to match the number of CPU cores available on the querier. Set it to slightly less than the number of cores if you are in a shared environment.
# If using the query front end, then run enough queriers to match: `(querier.worker-parallelism * number of query front end replicas) / querier.max-concurrent`.
# Without enough queriers, queries may queue in the query front end, which will affect the quality of service.
```

File: pkg/querier/querier.go [hunk 4]
```diff
@@ -56,7 +56,7 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
...
 	f.DurationVar(&cfg.ExtraQueryDelay, "querier.extra-query-delay", 0, "Time to wait before sending more than the minimum successful query requests.")
 	f.DurationVar(&cfg.QueryIngestersWithin, "querier.query-ingesters-within", 0, "Maximum lookback beyond which queries are not sent to ingester. 0 means all queries are sent to ingester.")
-	f.IntVar(&cfg.MaxConcurrent, "querier.max-concurrent", 20, "The maximum number of concurrent queries.")
+	f.IntVar(&cfg.MaxConcurrent, "querier.max-concurrent", 20, `The maximum number of top-level LogQL queries that will execute at the same time, per querier process. It should be set to match the number of CPU cores available on the querier (or slightly less than the number of cores if you are in a shared environment). If using the query frontend, Then you should run enough queriers to match ('-querier.worker-parallelism' * 'number of query frontend replicas')/'-querier.max-concurrent'. Otherwise queries may queue in the query frontend, which will affect QoS. Alternatively, consider using '-querier.worker-match-max-concurrent' to force worker parallelism to match '-querier.max-concurrent'`)
```
Comment (KMiller-Grafana) [inline pkg/querier/querier.go:59]: Can you make the prose here match the changes I put into lines 264-266?
Also, why did you change the double quotes delimiting the string into single back ticks? Looks wrong to me.

**Review by KMiller-Grafana — APPROVED** [review]
With some suggested doc changes, I think the doc portion of this looks good.

File: docs/sources/configuration/_index.md (hunk 1 above)
Comment (kavirajk) [reply docs/sources/configuration/_index.md]: @slim-bean Updated the doc!

File: docs/sources/configuration/_index.md (hunk 3 above)
Comment (kavirajk) [reply docs/sources/configuration/_index.md]: Thanks @KMiller-Grafana for the feedback. 
1. The minus sign(`-`) is actually part of the CLI flag name (e.g: `-querier.worker-paralleism` is the full name you see when you work with LogCLI. So we use it the same here.
2.  I would prefer to keep `frontend` rather than `front end`. Because its name of the component in Loki. (`Query frontend`

I agree with other changes :) Thanks!

File: pkg/querier/querier.go (hunk 4 above)
Comment (kavirajk) [reply pkg/querier/querier.go]: 👍

**Review by cyriltovena — APPROVED** [review]
LGTM

File: docs/sources/configuration/_index.md [hunk 5]
```diff
@@ -260,7 +260,11 @@ The `querier_config` block configures the Loki Querier.
 # CLI flag: -querier.query-ingesters-within
 [query_ingesters_within: <duration> | default = 0s]
 
-# The maximum number of concurrent queries allowed.
+# The maximum number of top-level LogQL queries that will execute at the same time, per querier process.
+# Set it to match the number of CPU cores available on the querier. Set it to slightly less than the number of cores if you are in a shared environment.
+# If using the query frontend, then you should run enough queriers to match (`-querier.worker-parallelism` * `number of query frontend replicas`)/`-querier.max-concurrent`.
+# Without enough queriers, queries may queue in the query frontend, which will affect the quality of service.
+# Alternatively, consider using `-querier.worker-match-max-concurrent` to force worker parallelism to match `-querier.max-concurrent`.
```
Comment (owen-d) [inline docs/sources/configuration/_index.md:267]:
```suggestion
# If using the query frontend, this should be set to at least (-querier.worker-parallelism * number of query frontend replicas).
# Otherwise queries may queue in the queriers and not the frontend, which will affect QoS.
# Alternatively, consider using -querier.worker-match-max-concurrent to force worker parallelism to match -querier.max-concurrent.
```

File: pkg/querier/querier.go [hunk 6]
```diff
@@ -56,7 +56,7 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
...
 	f.DurationVar(&cfg.ExtraQueryDelay, "querier.extra-query-delay", 0, "Time to wait before sending more than the minimum successful query requests.")
 	f.DurationVar(&cfg.QueryIngestersWithin, "querier.query-ingesters-within", 0, "Maximum lookback beyond which queries are not sent to ingester. 0 means all queries are sent to ingester.")
-	f.IntVar(&cfg.MaxConcurrent, "querier.max-concurrent", 20, "The maximum number of concurrent queries.")
+	f.IntVar(&cfg.MaxConcurrent, "querier.max-concurrent", 20, "The maximum number of top-level LogQL queries that will execute at the same time, per querier process. Set it to match the number of CPU cores available on the querier. Set it to slightly less than the number of cores if you are in a shared environment. If using the query frontend, then you should run enough queriers to match ('-querier.worker-parallelism' * 'number of query frontend replicas')/'-querier.max-concurrent'. Without enough queriers, queries may queue in the query frontend, which will affect the quality of service. Alternatively, consider using '-querier.worker-match-max-concurrent' to force worker parallelism to match '-querier.max-concurrent'")
```
Comment (owen-d) [inline pkg/querier/querier.go:59]: Please update this with the new explanation.

**Review by owen-d — COMMENTED** [review]
A small adjustment then LGTM
